File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        post = nums[len(nums)-1]
        answer = [1 for i in range(len(nums))]
        
        for i in range(1, len(nums)):
            answer[i] *= nums[i-1] * answer[i] * answer[i-1]
            
            
        for i in range(len(nums) - 2, -1, -1):
            answer[i] *= post
            post *= nums[i]
            
        return answer

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        int[] postfix = new int[n];
        int[] result = new int[n];
        prefix[0] = 1;
        postfix[n-1] = 1;
        for(int i = 0; i<n-1; i++) {
            prefix[i+1] = prefix[i] * nums[i];
        }
        for(int i = n - 2; i>=0; i--) {
            postfix[i] = postfix[i + 1] * nums[i + 1];
        }
        
        for(int i = 0; i<n; i++) {
            result[i] = prefix[i] * postfix[i];
        }
        return result;
    }
}